builtin/logical/pki/path_config_ca.go

```diff
@@ -13,24 +16,443 @@ func pathConfigCA(b *backend) *framework.Path {
+			`The "exported" path parameter must be "internal" or "exported"`)
+		return
+	}
+
+	format = data.Get("format").(string)
+	switch format {
+	case "pem":
+	case "der":
+	default:
+		errorResp = logical.ErrorResponse(
+			`The "format" path parameter must be "pem" or "der"`)
+		return
+	}
+
+	role = &roleEntry{
+		TTL:              data.Get("ttl").(string),
+		KeyType:          data.Get("key_type").(string),
+		KeyBits:          data.Get("key_bits").(int),
+		AllowLocalhost:   true,
+		AllowAnyName:     true,
+		AllowIPSANs:      true,
+		EnforceHostnames: false,
+	}
+
+	switch role.KeyType {
+	case "rsa":
+		switch role.KeyBits {
+		case 1024:
+		case 2048:
+		case 4096:
+		case 8192:
+		default:
+			errorResp = logical.ErrorResponse(fmt.Sprintf(
+				"unsupported bit length for RSA key: %d", role.KeyBits))
+			return
+		}
+	case "ec":
+		switch role.KeyBits {
+		case 224:
+		case 256:
+		case 384:
+		case 521:
+		default:
+			errorResp = logical.ErrorResponse(fmt.Sprintf(
+				"unsupported bit length for EC key: %d", role.KeyBits))
+			return
+		}
+	default:
+		errorResp = logical.ErrorResponse(fmt.Sprintf(
+			"unknown key type %s", role.KeyType))
+		return
+	}
+
+	return
+}
+
+func (b *backend) pathCAGenerateRoot(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	exported, format, role, errorResp := b.getGenerationParams(data)
+	if errorResp != nil {
+		return errorResp, nil
+	}
+
+	maxPathLengthIface, ok := data.GetOk("max_path_length")
+	if ok {
+		maxPathLength := maxPathLengthIface.(int)
+		role.MaxPathLength = &maxPathLength
+	}
+
+	var resp *logical.Response
+	parsedBundle, err := generateCert(b, role, nil, true, req, data)
+	if err != nil {
+		switch err.(type) {
+		case certutil.UserError:
+			return logical.ErrorResponse(err.Error()), nil
+		case certutil.InternalError:
+			return nil, err
+		}
+	}
+
+	cb, err := parsedBundle.ToCertBundle()
+	if err != nil {
+		return nil, fmt.Errorf("error converting raw cert bundle to cert bundle: %s", err)
+	}
+
+	resp = &logical.Response{
+		Data: map[string]interface{}{
+			"serial_number": cb.SerialNumber,
+			"expiration":    int64(parsedBundle.Certificate.NotAfter.Unix()),
+		},
+	}
+
+	switch format {
+	case "pem":
+		resp.Data["certificate"] = cb.Certificate
+		resp.Data["issuing_ca"] = cb.IssuingCA
+		if exported {
+			resp.Data["private_key"] = cb.PrivateKey
+			resp.Data["private_key_type"] = cb.PrivateKeyType
+		}
+	case "der":
+		resp.Data["certificate"] = base64.StdEncoding.EncodeToString(parsedBundle.CertificateBytes)
+		resp.Data["issuing_ca"] = base64.StdEncoding.EncodeToString(parsedBundle.IssuingCABytes)
+		if exported {
+			resp.Data["private_key"] = base64.StdEncoding.EncodeToString(parsedBundle.PrivateKeyBytes)
+			resp.Data["private_key_type"] = cb.PrivateKeyType
+		}
+	}
+
+	entry, err := logical.StorageEntryJSON("config/ca_bundle", cb)
+	if err != nil {
+		return nil, err
+	}
+	err = req.Storage.Put(entry)
```

Comment:
I wonder if we should validate or guard against multiple generations of a CA for a given backend. Seeing as it will blow away the existing cert, could be a pretty awkward situation.